### PR TITLE
Implement Star set

### PR DESCRIPTION
**Pull request by mforets**
Implement the new set representation from https://github.com/JuliaReach/LazySets.jl/pull/2132; other methods to be added in subsequent PRs.

joint with @SebastianGuadalupe

### CODE REVIEW

**Comment by mforets** [comment]
cc @castrong

we're splitting https://github.com/JuliaReach/LazySets.jl/pull/2132 (and what's left to be done there) into smaller chunks because it's more manageable that way :hocho: :cheese:

File: docs/src/lib/interfaces.md [hunk 1]
```diff
@@ -457,3 +457,15 @@ linear_map(::AbstractMatrix, ::AbstractAffineMap)
...
 * [Reset map (ResetMap)](@ref def_ResetMap)
 * [Translation](@ref def_Translation)
+
+## [Star sets (AbstractStar)](@id def_AbstractStar)
+
+An affine map consists of a linear map and a translation.
+
+```@docs
+AbstractAffineMap
+```
```
Comment (schillic) [inline docs/src/lib/interfaces.md:467]:
````suggestion
An affine map consists of a linear map and a translation.

```@docs
AbstractAffineMap
```

## [Star sets (AbstractStar)](@id def_AbstractStar)

````

File: src/Interfaces/AbstractStar.jl [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+Abstract supertype for all star set types.
+
+### Notes
+
+A set ``X`` is star-like (also known as generalized star) if it can be represented
+by a center ``x₀ ∈ \\mathbb{R}^n`` and ``m`` vectors ``v₁, …, vₘ`` forming the basis,
+and a predicate ``P : \\mathbb{R}^n → \\{⊤, ⊥\\}`` such that
+
+```math
+    X = \\{x ∈ \\mathbb{R}^n : x = x₀ + \\sum_{i=1}^m α_i v_i,,~~\\textrm{s.t. } P(α) = ⊤ \\}.
```
Comment (schillic) [inline src/Interfaces/AbstractStar.jl:13]: What are `α_i` and `α`?

File: src/Sets/Star.jl [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+This example is drawn from Example 1 in [2]. Consider the two-dimensional plane
+``\\mathbb{R}^2``. Let
+
+```jldoctest star_constructor
+julia> V = [[1.0, 0.0], [0.0, 1.0]]
+2-element Array{Array{Float64,1},1}:
+ [1.0, 0.0]
+ [0.0, 1.0]
+```
+be the bais vectors and take
```
Comment (schillic) [inline src/Sets/Star.jl:44]:
```suggestion
be the basis vectors and take
```

File: src/Sets/Star.jl [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+### Examples
+
+This example is drawn from Example 1 in [2]. Consider the two-dimensional plane
+``\\mathbb{R}^2``. Let
+
+```jldoctest star_constructor
+julia> V = [[1.0, 0.0], [0.0, 1.0]]
+2-element Array{Array{Float64,1},1}:
+ [1.0, 0.0]
+ [0.0, 1.0]
```
Comment (schillic) [inline src/Sets/Star.jl:42]: I would suggest to not print such trivial intermediate results. They are easy to read from the statements and just clutter the text.

File: src/Sets/Star.jl [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+```math
+ Θ = \\{(x, y) ∈ \\mathbb{R}^2 : (2 ≤ x ≤ 4) ∧ (2 ≤ y ≤ 4) \\}
+```
+We can check that assertion:
+
+```julia
+julia> Θ = Hyperrectangle(low=[2, 2.], high=[4, 4.])
+Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}([3.0, 3.0], [1.0, 1.0])
+
+julia> isequivalent(X, Θ) # FIXME
```
Comment (schillic) [inline src/Sets/Star.jl:94]: Something missing here? Also, `Θ` looks like a 0 in the code. Maybe just use `Y`?

File: src/Sets/Star.jl [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    c::VN # center
+    V::MN # basis
+    P::PT # predicate
+
+    # default constructor with size checks
+    function Star(c::VN, V::MN, P::PT) where {N, VN<:AbstractVector{N}, MN<:AbstractMatrix{N}, PT<:AbstractPolyhedron{N}}
+        @assert length(c) == size(V, 1) "the center of the basis vectors should be compatible, " *
+                                        "but they are of length $(length(c)) and $(size(V, 1)) respectively"
+
+        @assert dim(P) == size(V, 2) "the number of basis vectors should be compatible, " *
```
Comment (schillic) [inline src/Sets/Star.jl:125]:
```suggestion
        @assert dim(P) == size(V, 2) "the number of basis vectors should be compatible " *
```

File: src/Sets/Star.jl [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+end
+
+
+# constructor from center and list of generators
+function Star(c::VN, Vlist::AbstractVector{VN}, P::PT) where {N, VN<:AbstractVector{N}, PT<:AbstractPolyhedron{N}}
+    MT = _matrix_type(VN)
+    V = MT(undef, length(c), length(Vlist))
+    for (j, vj) in enumerate(Vlist)
+        @inbounds V[:, j] = vj
+    end
```
Comment (schillic) [inline src/Sets/Star.jl:139]: This could become a helper function `_to_matrix`.

File: src/Sets/Star.jl (hunk 5 above)
Comment (mforets) [reply src/Sets/Star.jl:94]: this requires some prior work in `isequivalent` and we need to add an inclusion check for stars. but i can make a new issue out of this  comment, that's probably a better approach.

File: src/Interfaces/AbstractStar.jl (hunk 2 above)
Comment (mforets) [reply src/Interfaces/AbstractStar.jl:13]: i've extended the statement, thanks for checking

**Review by schillic — APPROVED** [review]

**Pull request merged**